osu.Game/Screens/Play/MasterGameplayClockContainer.cs

```diff
@@ -0,0 +1,233 @@
+        {
+            // Lazer's audio timings in general doesn't match stable. This is the result of user testing, albeit limited.
+            // This only seems to be required on windows. We need to eventually figure out why, with a bit of luck.
+            platformOffsetClock = new HardwareCorrectionOffsetClock(source) { Offset = RuntimeInfo.OS == RuntimeInfo.Platform.Windows ? 15 : 0 };
+
+            // the final usable gameplay clock with user-set offsets applied.
+            userOffsetClock = new HardwareCorrectionOffsetClock(platformOffsetClock);
+
+            return localGameplayClock = new LocalGameplayClock(userOffsetClock);
+        }
+
+        /// <summary>
+        /// Changes the backing clock to avoid using the originally provided track.
+        /// </summary>
+        public void StopUsingBeatmapClock()
+        {
+            removeSourceClockAdjustments();
+            ChangeSource(new TrackVirtual(beatmap.Track.Length));
+            addSourceClockAdjustments();
+        }
+
+        private bool speedAdjustmentsApplied;
+
+        private void addSourceClockAdjustments()
+        {
+            if (speedAdjustmentsApplied)
+                return;
+
+            Track.AddAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
+            Track.AddAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);
+
+            localGameplayClock.MutableNonGameplayAdjustments.Add(pauseFreqAdjust);
+            localGameplayClock.MutableNonGameplayAdjustments.Add(UserPlaybackRate);
+
+            speedAdjustmentsApplied = true;
+        }
+
+        private void removeSourceClockAdjustments()
+        {
+            if (!speedAdjustmentsApplied)
+                return;
+
+            Track.RemoveAdjustment(AdjustableProperty.Frequency, pauseFreqAdjust);
+            Track.RemoveAdjustment(AdjustableProperty.Tempo, UserPlaybackRate);
+
+            localGameplayClock.MutableNonGameplayAdjustments.Remove(pauseFreqAdjust);
+            localGameplayClock.MutableNonGameplayAdjustments.Remove(UserPlaybackRate);
+
+            speedAdjustmentsApplied = false;
+        }
+
+        protected override void Dispose(bool isDisposing)
+        {
+            base.Dispose(isDisposing);
+            removeSourceClockAdjustments();
+        }
+
+        private class HardwareCorrectionOffsetClock : FramedOffsetClock
+        {
+            // we always want to apply the same real-time offset, so it should be adjusted by the difference in playback rate (from realtime) to achieve this.
+            // base implementation already adds offset at 1.0 rate, so we only add the difference from that here.
+            public override double CurrentTime => base.CurrentTime + Offset * (Rate - 1);
+
+            public HardwareCorrectionOffsetClock(IClock source, bool processSource = true)
+                : base(source, processSource)
+            {
+            }
+        }
+
+        private class LocalGameplayClock : GameplayClock
```

Comment:
Wondering if this should be called `MasterGameplayClock` instead, now that it exists in this class? the "Local" term had me a bit confused at first.